awss3/store.go

```diff
@@ -0,0 +1,703 @@
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/pborman/uuid"
+	"golang.org/x/net/context"
+
+	"github.com/lytics/cloudstorage"
+)
+
+const (
+	// StoreType = "s3" this is used to define the storage type to create
+	// from cloudstorage.NewStore(config)
+	StoreType = "s3"
+
+	// Configuration Keys.  These are the names of keys
+	// to look for in the json map[string]string to extract for config.
+
+	// ConfKeyAccessKey config key name of the aws access_key(id) for auth
+	ConfKeyAccessKey = "access_key"
+	// ConfKeyAccessSecret config key name of the aws acccess secret
+	ConfKeyAccessSecret = "access_secret"
+	// ConfKeyARN config key name of the aws ARN name of user
+	ConfKeyARN = "arn"
+	// ConfKeyDisableSSL config key name of disabling ssl flag
+	ConfKeyDisableSSL = "disable_ssl"
+	// Authentication Source's
+
+	// AuthAccessKey is for using aws access key/secret pairs
+	AuthAccessKey cloudstorage.AuthMethod = "aws_access_key"
+)
+
+var (
+	// Retries number of times to retry upon failures.
+	Retries = 55
+	// PageSize is default page size
+	PageSize = 2000
+
+	// ErrNoS3Session no valid session
+	ErrNoS3Session = fmt.Errorf("no valid aws session was created")
+	// ErrNoAccessKey error for no access_key
+	ErrNoAccessKey = fmt.Errorf("no settings.access_key")
+	// ErrNoAccessSecret error for no settings.access_secret
+	ErrNoAccessSecret = fmt.Errorf("no settings.access_secret")
+	// ErrNoAuth error for no findable auth
+	ErrNoAuth = fmt.Errorf("No auth provided")
+)
+
+func init() {
+	// Register this Driver (s3) in cloudstorage driver registry.
+	cloudstorage.Register(StoreType, func(conf *cloudstorage.Config) (cloudstorage.Store, error) {
+		client, sess, err := NewClient(conf)
+		if err != nil {
+			return nil, err
+		}
+		return NewStore(client, sess, conf)
+	})
+}
+
+type (
+	// FS Simple wrapper for accessing s3 files, it doesn't currently implement a
+	// Reader/Writer interface so not useful for stream reading of large files yet.
+	FS struct {
+		PageSize  int
+		ID        string
+		client    *s3.S3
+		sess      *session.Session
+		endpoint  string
+		bucket    string
+		cachepath string
+	}
+
+	object struct {
+		fs         *FS
+		o          *s3.GetObjectOutput
+		cachedcopy *os.File
+
+		name      string    // aka "key" in s3
+		updated   time.Time // LastModifyied in s3
+		metadata  map[string]string
+		bucket    string
+		readonly  bool
+		opened    bool
+		cachepath string
+
+		infoOnce sync.Once
+		infoErr  error
+	}
+)
+
+// NewClient create new AWS s3 Client.  Uses cloudstorage.Config to read
+// necessary config settings such as bucket, region, auth.
+func NewClient(conf *cloudstorage.Config) (*s3.S3, *session.Session, error) {
+
+	awsConf := aws.NewConfig().
+		WithHTTPClient(http.DefaultClient).
+		WithMaxRetries(aws.UseServiceDefaultRetries).
+		WithLogger(aws.NewDefaultLogger()).
+		WithLogLevel(aws.LogOff).
+		WithSleepDelay(time.Sleep)
```

Comment:
lol they let you override the sleep func used, this must be exposed for testing???

Reply:
or throttling?